Boot mem handling

jenswi-linaro opened the pull request:
This pull request handles previously unused memory during boot—currently mostly temporary memory allocations, but that will be extended in the next pull requests.

The commit "core: mm: allocate temporary memory map array"  demonstrates this by using dynamic allocation instead of hard-coded allocation of `static_mmap_regions[]`.

jforissier reviewed (commented):
For "mk/clang.mk: -Wno-gnu-alignof-expression":

s/dereferensing/dereferencing/
s/CLANG/Clang/

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
- For "core: arm64: increase thread stack size for debug":
  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
- For "core: mm: add vaddr_to_phys()":
  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
- For "core: remove CORE_MEM_TA_RAM":
  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
OK to squash in the fixes and move up the last commits a bit?

jforissier commented:
> OK to squash in the fixes and move up the last commits a bit?

Fine with me

jenswi-linaro commented:
Tags applied, squashed in fixes, and moved the last added commits a bit earlier to fix problems before they occur.

jforissier reviewed (commented):
- For "core: mm: allow unmapping VCORE_FREE":
  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -489,17 +489,20 @@ __vcore_nex_rw_end = __vcore_nex_rw_start + __vcore_nex_rw_size;
 #endif /* CFG_CORE_RODATA_NOEXEC */
 __vcore_init_rx_end = __vcore_init_rx_start + __vcore_init_rx_size;
 __vcore_init_ro_end = __vcore_init_ro_start + __vcore_init_ro_size;
+ASSERT(__vcore_init_ro_start == __vcore_init_ro_end,
+       "__vcore_init_ro_start should follow __vcore_init_ro_end")
```
s/__vcore_init_ro_end/__vcore_init_rx_end/? And in the commit description "after __vcore_init_rx_end"?

jenswi-linaro replied:
You're right, I'll update the commit.

jforissier reviewed (commented):
Comment on "core: arm,pager: make __vcore_init_ro_start follow __vcore_init_rx_end".

jenswi-linaro commented:
Force push to update "core: arm,pager: make __vcore_init_ro_start follow __vcore_init_rx_end" and applied the tags.

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -2639,10 +2648,48 @@ void core_mmu_init_phys_mem(void)
+#ifdef TEE_SDP_TEST_MEM_BASE
+		if (TEE_SDP_TEST_MEM_SIZE) {
+			pa = vaddr_to_phys(TEE_SDP_TEST_MEM_BASE);
+			carve_out_core_mem(pa, pa + TEE_SDP_TEST_MEM_BASE);
```
s/TEE_SDP_TEST_MEM_BASE/TEE_SDP_TEST_MEM_SIZE/

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -2623,6 +2623,15 @@ static TEE_Result teecore_init_pub_ram(void)
+{
+	tee_mm_entry_t *mm __maybe_unused = NULL;
+
+	DMSG("%#"PRIxPA" .. %#"PRIxPASZ, pa, end_pa);
```
s/PRIxPASZ/PRIxPA/

etienne-lms reviewed (commented):
Very minor review comments.

I was looking at the CFG_BOOT_MEM part and experienced the patches on my Armv7 boards. Commit "core: arm: enable CFG_BOOT_MEM unconditionally" makes my platforms to segfault at primary core init (only when pager is disabled). I dumped the traces below, in case it helps:

```
I/TC: Early console on UART#4
D/TC:0   add_phys_mem:750 VCORE_UNPG_RX_PA type TEE_RAM_RX 0xde000000 size 0x00076000
D/TC:0   add_phys_mem:750 VCORE_UNPG_RW_PA type TEE_RAM_RW 0xde076000 size 0x0005a000
D/TC:0   add_phys_mem:750 VCORE_UNPG_RW_PA type SEC_RAM_OVERALL 0xde076000 size 0x0005a000
D/TC:0   add_phys_mem:750 VCORE_FREE_PA type TEE_RAM_RW 0xde0d0000 size 0x00130000
D/TC:0   merge_mmaps:703 Merging 0xde076000..0xde0cffff and 0xde0d0000..0xde1fffff
D/TC:0   add_phys_mem:750 VCORE_FREE_PA type SEC_RAM_OVERALL 0xde0d0000 size 0x00130000
D/TC:0   merge_mmaps:703 Merging 0xde076000..0xde0cffff and 0xde0d0000..0xde1fffff
D/TC:0   add_phys_mem:750 next_pa type SEC_RAM_OVERALL 0xde200000 size 0x01e00000
D/TC:0   merge_mmaps:703 Merging 0xde076000..0xde1fffff and 0xde200000..0xdfffffff
D/TC:0   add_phys_mem:750 GIC_BASE type IO_SEC 0xa0000000 size 0x00200000
...
D/TC:0   add_phys_mem:750 APB1_BASE type IO_NSEC 0x40000000 size 0x00200000
D/TC:0   add_va_space:794 type RES_VASPACE size 0x00a00000
D/TC:0   add_va_space:794 type SHM_VASPACE size 0x02000000
D/TC:0   dump_mmap_table:920 type SEC_RAM_OVERALL va 0xd76ec000..0xd9675fff pa 0xde076000..0xdfffffff size 0x01f8a000 (pgdir)
D/TC:0   dump_mmap_table:920 type IO_SEC       va 0xd9800000..0xd99fffff pa 0xa0000000..0xa01fffff size 0x00200000 (pgdir)
...
D/TC:0   dump_mmap_table:920 type IO_NSEC      va 0xdb000000..0xdb1fffff pa 0x40000000..0x401fffff size 0x00200000 (pgdir)
D/TC:0   dump_mmap_table:920 type RES_VASPACE  va 0xdb200000..0xdbbfffff pa 0x00000000..0x009fffff size 0x00a00000 (pgdir)
D/TC:0   dump_mmap_table:920 type SHM_VASPACE  va 0xdbe00000..0xdddfffff pa 0x00000000..0x01ffffff size 0x02000000 (pgdir)
D/TC:0   dump_mmap_table:920 type TEE_RAM_RX   va 0xde000000..0xde075fff pa 0xde000000..0xde075fff size 0x00076000 (smallpg)
D/TC:0   dump_mmap_table:920 type TEE_RAM_RW   va 0xde076000..0xde1fffff pa 0xde076000..0xde1fffff size 0x0018a000 (smallpg)
D/TC:0   core_mmu_xlat_table_alloc:527 xlat tables used 1 / 4
D/TC:0   core_mmu_xlat_table_alloc:527 xlat tables used 2 / 4
D/TC:0   core_mmu_xlat_table_alloc:527 xlat tables used 3 / 4
D/TC:0   carve_out_core_mem:2654 0xde000000 .. 0xde200000
D/TC:0   boot_mem_release_unused:152 Allocated 0 bytes at va 0xde0d0000 pa 0xde0d0000
D/TC:0   boot_mem_release_unused:156 Tempalloc 4380 bytes at va 0xde1feee4
D/TC:0   boot_mem_release_unused:178 Carving out 0xde000000..0xde0cffff
D/TC:0   boot_mem_release_unused:187 Releasing 1236992 bytes from va 0xde0d0000
I/TC: 
I/TC: Embedded DTB found
I/TC: OP-TEE version: 4.3.0-141-g0e3fb1edc (gcc version 11.3.1 20220712 (Arm GNU Toolchain 11.3.Rel1)) #36 Thu Sep 19 07:43:17 UTC 2024 arm
...
I/TC: Primary CPU initializing
...
D/TC:0 0 boot_mem_release_tmp_alloc:227 Releasing 8192 bytes from va 0xde1fe000
...
I/TC: Primary CPU switching to normal world boot
E/TC:0 0 
E/TC:0 0 Core data-abort at address 0xde0d0000 (translation fault)
E/TC:0 0  fsr 0x00002a07  ttbr0 0xde0ba000  ttbr1 0x00000000  cidr 0x0
E/TC:0 0  cpu #0          cpsr 0x800001d3
E/TC:0 0  r0 0xde0d0000      r4 0x2fff2d84    r8 0xde078828   r12 0x00000000
E/TC:0 0  r1 0xde1ff018      r5 0x00000000    r9 0xde0cf5c0    sp 0xde0cf5c0
E/TC:0 0  r2 0x00000040      r6 0xc0400000   r10 0xafaeadac    lr 0xde00021c
E/TC:0 0  r3 0x0000003f      r7 0x00000000   r11 0x00000000    pc 0xde000d24
E/TC:0 0 TEE load address @ 0xde000000
E/TC:0 0 Call stack:
E/TC:0 0  0xde000d24 dcache_cleaninv_range at core/arch/arm/kernel/cache_helpers_a32.S:53
E/TC:0 0  0xde00021c clear_bss at core/arch/arm/kernel/entry_a32.S:588
```
The data-abort occurs in `reset_primary` before switch to non-secure world, fluhsing vmem that is not mapped:
_core/arch/arm/kernel/entry_a32.S_:
```S
	/*
	 * In case we've touched memory that secondary CPUs will use before
	 * they have turned on their D-cache, clean and invalidate the
	 * D-cache before exiting to normal world.
	 */
588:	flush_cache_vrange(cached_mem_start, cached_mem_end)
```
Obviously not reproduced on vexpress-qemu_virt.

etienne-lms commented:
To sumup, the cache flush operation at _entry_a32.S:588_ above flushes a VA range (TEE RAM) that includes VA areas not mapped (released boot-mem areas IIUC), they trigger a data-abort exception in the CPU. Testing a bit, it seems Qemu (at least the tool/config we use) does not trap such accesses as illegals.

jenswi-linaro commented:
> To sumup, the cache flush operation at _entry_a32.S:588_ above flushes a VA range (TEE RAM) that includes VA areas not mapped (released boot-mem areas IIUC), they trigger a data-abort exception in the CPU. Testing a bit, it seems Qemu (at least the tool/config we use) does not trap such accesses as illegals.

Thanks for the detailed analysis, I hope the update fixes the problem.

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -2682,10 +2668,22 @@ static void __maybe_unused carve_out_core_mem(paddr_t pa, paddr_t end_pa)
+	if (IS_ENABLED(CFG_NS_VIRTUALIZATION)) {
+		paddr_t b1 = 0;
+		paddr_size_t s1 = 0;
 
```
> Commit "core: merge core_mmu_init_phys_mem() and core_mmu_init_virtualization()":
> 
> * Intentionally removed the info level trace `IMSG("Initializing virtualization support");` ?
>   (previously printed from _core/arch/arm/kernel/boot.c_)

Yes, I saw no point in keeping it.

> * inline comment above `... TA are loaded/executedNsec ...` is pasted from previous implementation but is is quite buggy. I think it's time to fix it.

Right, I'll remove it.

jforissier replied:
> > Commit "core: merge core_mmu_init_phys_mem() and core_mmu_init_virtualization()":
> > 
> > * Intentionally removed the info level trace `IMSG("Initializing virtualization support");` ?
> >   (previously printed from _core/arch/arm/kernel/boot.c_)
> 
> Yes, I saw no point in keeping it.

Do we have another way of knowing that the OP-TEE binary was compiled with virtualization support? Linux driver traces perhaps?

jenswi-linaro replied:
I've added "core: arm: add CFG_NS_VIRTUALIZATION boot log" to take care of that.

jenswi-linaro commented on core/arch/arm/arm.mk:
```diff
@@ -15,6 +15,10 @@ include mk/cc-option.mk
 CFG_CORE_TZSRAM_EMUL_SIZE ?= 458752
 
+# Mandatory for arch/arm to avoid ifdefs in the arch specific code
+CFG_BOOT_MEM=y
```
This one is needed or we'll get an error:
```
core/arch/arm/arm.mk:19: *** CFG_BOOT_MEM is set to 'n' (from file) but its value must be 'y'.  Stop.
```
This is due to the default value assigned in `mk/config.mk`

etienne-lms replied:
I think we should rather have in `mk/config.mk` something like:
```
ifeq ($(ARCH),arm)
$(call force,CFG_BOOT_MEM,y)
else
CFG_BOOT_MEM ?= n
endif

jenswi-linaro replied:
OK

jenswi-linaro commented:
Update, comments addressed.

jforissier commented:
For "core: arm: add CFG_NS_VIRTUALIZATION boot log":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -934,7 +936,15 @@ static void init_primary(unsigned long pageable_part, unsigned long nsec_entry)
 	core_mmu_init_phys_mem();
-	boot_mem_release_unused();
+	va = boot_mem_release_unused();
+	if (IS_ENABLED(CFG_ARM32_core) && !IS_ENABLED(CFG_WITH_PAGER)) {
```
You're right, I've pushed an update for that.

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -540,17 +536,17 @@ static void init_runtime(unsigned long pageable_part)
 	/*
-	 * Need physical memory pool initialized to be able to allocate
-	 * secure physical memory below.
+	 * The pager is about the be enabled below, evenual temporary boot
```
Thanks, I'll fix.

jenswi-linaro commented on core/include/kernel/linker.h:
```diff
@@ -46,6 +46,17 @@ extern const uint8_t __extab_end[];
+					  __vcore_free_start))
+#define VCORE_FREE_END_PA	((unsigned long)__vcore_free_end)
+#else
+#define VCORE_FREE_PA		PADDR_MAX
```
Sure, I'll add a comment.

jenswi-linaro commented:
Update. Let me know if I should squash in the updates.

etienne-lms commented:
I'm fine you squash the fixup commits.

jenswi-linaro commented:
Squashed in the updates.

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -66,7 +66,7 @@ static struct tee_mmap_region static_mmap_regions[CFG_MMAP_REGIONS
 						+ 1
 #endif
-						+ 1] __nex_bss;
+						+ 2] __nex_bss;
```
This change adds 1, 3 or 4 cells in `static_mmap_regions` (depending on `CFG_NS_VIRTUALIZATION` and `CFG_CORE_RWDATA_NOEXEC`).
Should this be reflected here? Or maybe add 4 cells, unconditionally to cover all cases? Or use explicit config cases (below, not that very helpful I guess)?
```patch
 static struct tee_mmap_region static_mmap_regions[CFG_MMAP_REGIONS
 #if defined(CFG_CORE_ASLR) || defined(CFG_CORE_PHYS_RELOCATABLE)
 						+ 1
 #endif
+#ifdef CFG_CORE_RWDATA_NOEXEC
+#ifdef CFG_NS_VIRTUALIZATION
+ 						+ 4
+#else
+						+ 3
+#endif
+#else
+						+ 1
+#endif /*CFG_CORE_RWDATA_NOEXEC*/
 						+ 1] __nex_bss;
```

jenswi-linaro replied:
Yeah, let's do that to keep things simple.

etienne-lms reviewed (commented):
Some comments on commit "core: mm: replace MEM_AREA_TA_RAM".

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -381,12 +381,8 @@ static TEE_Result mmap_clear_memtag(struct tee_mmap_region *map,
-	case MEM_AREA_NEX_RAM_RW:
-	case MEM_AREA_TEE_ASAN:
-	case MEM_AREA_TA_RAM:
+	case MEM_AREA_SEC_RAM_OVERALL:
```
> Asan memory is still referenced with `MEM_AREA_TEE_ASAN` memory area ID when `CFG_WITH_PAGER` is enabled. Won't this switch case need to be supported?

```
mk/config.mk:999: *** CFG_CORE_SANITIZE_KADDRESS and CFG_MEMTAG are not compatible.  Stop.
```
I don't think we need to care about `MEM_AREA_TEE_ASAN` since it's not used with memory tagging enabled.
 
> `MEM_AREA_TEE_RAM` is still used when `CFG_CORE_RWDATA_NOEXEC` is disabled. IS that an issue? Maybe it's a bit inconsistent to enable `CFG_MEMTAG` but not `CFG_CORE_RWDATA_NOEXEC`.

That shouldn't be a problem, everything should be covered by `MEM_AREA_SEC_RAM_OVERALL` nonetheless.

> Is is OK here to no more clear memtags for `MEM_AREA_TEE_RAM_RW` and `MEM_AREA_NEX_RAM_RW` areas?

The same here, is covered by `MEM_AREA_SEC_RAM_OVERALL`.

However, while looking at logs etc to double-check I noticed:
```
D/TC:0   dump_mmap_table:923 type SEC_RAM_OVERALL va 0x12100000..0x122fffff pa 0x0e100000..0x0e2fffff size 0x00200000 (pgdir)
...
D/TC:0   mmap_clear_memtag:392 Clearing tags for VA 0x12000000..0x121fffff
```

We may have this error in more than one place, see commit f01690c36528 ("core: fix mapping init debug trace") for the `dump_mmap_table()` print. I'll investigate further.

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -1081,12 +1080,22 @@ static void collect_mem_ranges(struct memory_map *mem_map)
+		 */
+		ADD_PHYS_MEM(MEM_AREA_TEE_RAM, ram_start,
+			     VCORE_UNPG_RX_PA - ram_start);
+		ADD_PHYS_MEM(MEM_AREA_SEC_RAM_OVERALL, ram_start,
```
Would it make more sense only to map the gap between `TEE_RAM_START` and `TEE_LOAD_ADDR` in `MEM_AREA_SEC_RAM_OVERALL`?

etienne-lms replied:
I think it would, unless some platforms load a TEE core image made on core binary image prepend with some data expected to be read-only mapped. Could be strange but current implementation allows that.
Maybe this old scheme could be deprecated. In that case, at least the commit message should explicitly state it IMHO.

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -1193,15 +1193,13 @@ static void assign_mem_granularity(struct memory_map *mem_map)
 		if (map_is_tee_ram(mem_map->map + n))
 			mem_map->map[n].region_size = SMALL_PAGE_SIZE;
+		else
+			mem_map->map[n].region_size = CORE_MMU_PGDIR_SIZE;
```
Previous implementation allowed to page map cacheable memory outside TEE RAM, (MEM_AREA_RAM_NSEC, MEM_AREA_RAM_SEC, MEM_AREA_ROM_SEC). I fear memory firewall permission violation on speculative accesses  for such memories, and likely few other corner issues on some platform specific memories.

jenswi-linaro replied:
This is the requested granularity. It will not mean mapping more memory than specified. So some ranges will be mapped by of mix of small-page and pgdir sized mappings (level 3 and level 2 with LPAE).

Take QEMU with hafnium as an example:
```
TEE_RAM_RWX  va 0x0e304000..0x0e503fff pa 0x0e304000..0x0e503fff size 0x00200000 (smallpg)
RES_VASPACE  va 0x0e600000..0x0effffff pa 0x00000000..0x009fffff size 0x00a00000 (pgdir)
SHM_VASPACE  va 0x0f000000..0x10ffffff pa 0x00000000..0x01ffffff size 0x02000000 (pgdir)
IO_SEC       va 0x11000000..0x1a080fff pa 0x09000000..0x12080fff size 0x09081000 (pgdir)
SEC_RAM_OVERALL va 0x1a300000..0x1affffff pa 0x0e300000..0x0effffff size 0x00d00000 (pgdir)
```
`IO_SEC` is then mapped as:
```
D/TC:0        [LVL2] VA:0x0011000000 PA:0x0009000000 DEV-RW-XN-S
D/TC:0        [LVL2] VA:0x0011200000 PA:0x0009200000 DEV-RW-XN-S
...
D/TC:0        [LVL2] VA:0x001a000000 TBL:0x000e3fa000 
D/TC:0          [LVL3] VA:0x001a000000 PA:0x0012000000 DEV-RW-XN-S
D/TC:0          [LVL3] VA:0x001a001000 PA:0x0012001000 DEV-RW-XN-S
...
D/TC:0          [LVL3] VA:0x001a080000 PA:0x0012080000 DEV-RW-XN-S
D/TC:0          [LVL3] VA:0x001a081000 INVALID
```
`SEC_RAM_OVERALL` as:
```
D/TC:0          [LVL3] VA:0x001a300000 PA:0x000e300000 MEM-RW-XN-S
D/TC:0          [LVL3] VA:0x001a301000 PA:0x000e301000 MEM-RW-XN-S
...
D/TC:0          [LVL3] VA:0x001a3ff000 PA:0x000e3ff000 MEM-RW-XN-S
D/TC:0        [LVL2] VA:0x001a400000 PA:0x000e400000 MEM-RW-XN-S
...
D/TC:0        [LVL2] VA:0x001ae00000 PA:0x000ee00000 MEM-RW-XN-S
D/TC:0        [LVL2] VA:0x001b000000 INVALID
```

So pgdir-sized `region_size` uses as few translation tables as possible, but it will not result in mapping beyond the requested range.

etienne-lms reviewed (commented):
Few review tags: `Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits:
"mk/clang.mk: -Wno-gnu-alignof-expression",
"core: arm64: increase thread stack size for debug",
"core: mm: add vaddr_to_phys()",
"core: remove CORE_MEM_TA_RAM",
"core: add VCORE_FREE_{PA,SZ,END_PA}",
"core: mm: allow unmapping VCORE_FREE",
"core: mm: unify secure core and TA memory",
"core: virt: phys_mem_core_alloc() use both pools",
"core: arm: core_mmu_v7.c: increase MAX_XLAT_TABLES by 2" and
"core: arm,pager: make __vcore_init_ro_start follow __vcore_init_rx_end"

A concern with commit "core: mm: map memory using requested block size".

jenswi-linaro commented:
Can I squash in the updates?

jenswi-linaro commented:
Tags applied

jenswi-linaro commented:
Update

jenswi-linaro commented:
Should I squash and rebase to continue the review?

etienne-lms commented:
Fine  with me. (sorry for the late feedback)

jenswi-linaro commented:
Squashed and rebased.

jenswi-linaro commented:
Reviews needed for:
- "core: mm: unify secure core and TA memory"
- "core: mm: map memory using requested block size"
- "core: mm,pager: map remaining physical memory"
- "core: add CFG_BOOT_MEM and boot_mem_*() functions"
- "core: arm: add boot_cached_mem_end"
- "core: arm: enable CFG_BOOT_MEM unconditionally"
- "core: mm: allocate temporary memory map array"
- "core: initialize guest physical memory early"
- "core: merge core_mmu_init_phys_mem() and core_mmu_init_virtualization()"

The commit "core: mm: allocate temporary memory map array" demonstrates the ridding of the hardcoded size of the array `static_mmap_regions`. I plan to do the same for the translation tables, number of cores, number of threads, and the heap. Much of the groundwork is already done, but this PR is needed before I can do the rest.

jenswi-linaro commented:
Ping?

etienne-lms commented:
FYI, i'm looking at it. Still struggling to make my ind clear on commit "core: mm: replace MEM_AREA_TA_RAM". I'll try to give some feedback soon.

etienne-lms reviewed (commented):
A tiny bit of progress in the review...

For commit "core: mm: replace MEM_AREA_TA_RAM": 2 comments otherwise looks good to me. `Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` with these 2 comments addressed or not (no effective functional changes).

`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: mm: map memory using requested block size",
"core: mm,pager: map remaining physical memory" and
"core: arm: add boot_cached_mem_end" (with `#ifdef` issue fixed).

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -1081,13 +1083,24 @@ static void collect_mem_ranges(struct memory_map *mem_map)
 			     VCORE_UNPG_RX_PA - ram_start);
+		assert(VCORE_UNPG_RX_PA >= ram_start);
+		tee_ram_initial_offs = VCORE_UNPG_RX_PA - ram_start;
+		DMSG("tee_ram_initial_offs %zx", tee_ram_initial_offs);
```
That makes sense.

jenswi-linaro commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -468,12 +470,9 @@ clear_nex_bss:
 END_FUNC _start
 DECLARE_KEEP_INIT _start
 
+#ifndef CFG_WITH_PAGER
```
Thanks, I'll fix.

jenswi-linaro commented on core/mm/core_mmu.c:
```diff
@@ -1096,41 +1109,51 @@ static void collect_mem_ranges(struct memory_map *mem_map)
 			ADD_PHYS_MEM(MEM_AREA_NEX_RAM_RO, VCORE_UNPG_RW_PA,
 				     VCORE_UNPG_RW_SZ);
+			ADD_PHYS_MEM(MEM_AREA_SEC_RAM_OVERALL, VCORE_UNPG_RW_PA,
+				     VCORE_UNPG_RW_SZ);
```
The idea is that MEM_AREA_SEC_RAM_OVERALL should cover all memory that doesn't have to be mapped RO or RX. It makes it easier when we know need to update data at a physical address, for instance when traversing and updating translation tables. All secure physical memory can be reached via a lookup with
`phys_to_virt(pa, MEM_AREA_SEC_RAM_OVERALL, sz)`

MEM_AREA_SEC_RAM_OVERALL isn't about the security level of the memory it's to have easy access via a present mapping. We should normally not save virtual pointers into MEM_AREA_SEC_RAM_OVERALL when it's an aliased mapping. We may need to make exceptions to that, but this is a starting point.

jenswi-linaro replied:
So I'd rather not complicate things more for the pager configuration.

jenswi-linaro commented:
Comments addressed (except for the MEM_AREA_SEC_RAM_OVERALL comment) and tags applied.

etienne-lms commented on core/include/kernel/boot.h:
```diff
@@ -105,4 +105,23 @@ void discover_nsec_memory(void);
 /* Add reserved memory for static shared memory in the device-tree */
 int mark_static_shm_as_reserved(struct dt_descriptor *dt);
 
+#ifdef CFG_BOOT_MEM
```
A bit of description could help. Maybe something like what's in the commit message:
```
Stack-like memory allocations during boot before a heap has
been configured. Boot_mem heap considers possible relocation
during virtual memory initialization. Unused memory is unmapped
and released to SECRAM_OVERALL allocation pool once MMU is
initialized.
```
(if I understood correctly)

jenswi-linaro replied:
Thanks, I'll add something like that.

etienne-lms reviewed (commented):
Few comments on commit "core: add CFG_BOOT_MEM and boot_mem_*() functions".

jenswi-linaro commented on core/mm/boot_mem.c:
```diff
@@ -0,0 +1,232 @@
+	struct boot_mem_reloc *next;
+};
+
+struct boot_mem_desc {
```
I'm adding it, thanks.

jenswi-linaro commented on core/mm/boot_mem.c:
```diff
@@ -0,0 +1,232 @@
+	}
+}
+
+void *boot_mem_alloc(size_t len, size_t align)
```
Yes, that function will be needed in later patches to allocate translation tables, stacks, etc.

jenswi-linaro commented:
Updated

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: add CFG_BOOT_MEM and boot_mem_*() functions".

jenswi-linaro commented:
Squashed and tag applied.

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -1583,7 +1620,16 @@ void __weak core_init_mmu_map(unsigned long seed, struct core_mmu_config *cfg)
 
-	mem_map = static_memory_map;
+	if (IS_ENABLED(CFG_BOOT_MEM)) {
+		mem_map.alloc_count = MEM_AREA_MAXTYPE;
```
s/MEM_AREA_MAXTYPE/CFG_MMAP_REGIONS/

etienne-lms commented on core/include/mm/core_mmu.h:
```diff
@@ -295,6 +295,7 @@ extern const unsigned long core_mmu_tee_load_pa;
 
 void core_init_mmu_map(unsigned long seed, struct core_mmu_config *cfg);
 void core_init_mmu_regs(struct core_mmu_config *cfg);
+void core_mmu_save_mem_map(void);
```
Maybee add a short description: `Copy static memory map from boot_mem to heap when CFG_BOOT_MEM is enabled`

etienne-lms commented on core/mm/core_mmu.c:
```diff
@@ -2691,10 +2677,18 @@ static void __maybe_unused carve_out_core_mem(paddr_t pa, paddr_t end_pa)
 
-	if (!IS_ENABLED(CFG_NS_VIRTUALIZATION)) {
+		static_assert(ARRAY_SIZE(secure_only) <= 2);
+		if (ARRAY_SIZE(secure_only) == 2) {
```
An an empty line above.

+ I think it would be worth restoring the info level trace message
`IMSG("Initializing virtualization support");`

jenswi-linaro replied:
Sure, I'll add an empty line.

We don't want it in this file. We're adding a new debug level message in "core: arm: add CFG_NS_VIRTUALIZATION boot log". Do you mean this print is important enough that the level should be info rather than debug? Debug level is the right level IMO.

etienne-lms replied:
No, i'm fine if the old trace message is  removed.

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: arm: enable CFG_BOOT_MEM unconditionally" with a minor comment.

Comments on commit "core: mm: allocate temporary memory map array",
otherwise LGTM: `Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` with the `alloc_count` issue addressed (in case my fix proposal is valid).

`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"core: initialize guest physical memory early",
"core: merge core_mmu_init_phys_mem() and core_mmu_init_virtualization()"  (with minor comments addressed) and
"core: arm: add CFG_NS_VIRTUALIZATION boot log".

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -940,14 +917,55 @@ static void init_primary(unsigned long pageable_part, unsigned long nsec_entry)
+		boot_cached_mem_end = va;
+	}
+
+	IMSG_RAW("\n");
```
OK

jenswi-linaro commented:
Addressed comments and added tags, except for "core: merge core_mmu_init_phys_mem() and core_mmu_init_virtualization()". Please let me know if anything more is needed for that commit.

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: merge core_mmu_init_phys_mem() and core_mmu_init_virtualization()".

jenswi-linaro commented:
Tag applied. Rebased on master to resolve the conflict. All commits are reviewed. Thanks a lot for the help.
